keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+   validated against them in admission. This category, along with the admission controller, will be
+   addressed in a subsequent update to this proposal.
+3. **Informational** - These fields surface semantic properties of the runtime to guide runtime
+   selection. Examples could include hardware virtualization, stability, and I/O performance
+   class. These fields are out-of-scope for the initial proposal.
+
+#### API Specification
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+type RuntimeClass struct {
+    metav1.TypeMeta
+    // ObjectMeta minimally includes the RuntimeClass name, which is used to reference the class.
+    // Namespace should be left blank.
+    metav1.ObjectMeta
+
+    Spec RuntimeClassSpec
+}
+
+type RuntimeClassSpec struct {
+    // Parameters holds opaque parameters that are interpreted by the runtime (CRI implementation).
+    Parameters map[string]string
+}
+```
+
+The runtime is selected by the pod by specifying the RuntimeClass in the PodSpec:
+
+```go
+type PodSpec struct {
+    ...
+    // RuntimeClassName refers to a RuntimeClass object with the same name,
+    // which should be used to run this pod.
+    // +optional
+    RuntimeClassName string
+    ...
+}
+```
+
+The `legacy` RuntimeClass name is reserved. The legacy RuntimeClass is defined to be fully backwards
+compatible with current Kubernetes. This means that the legacy runtime does not specify any runtime
+parameters or perform any feature validation (all features are "supported").
+
+```go
+const (
+    // RuntimeClassNameLegacy is a reserved RuntimeClass name. The legacy
+    // RuntimeClass does not specify any runtime parameters or perform any
+    // feature validation.
+    RuntimeClassNameLegacy = "legacy"
+)
+```
+
+An unspecified RuntimeClassName `""` is equivalent to the `legacy` RuntimeClass, though the field is
+not defaulted to `legacy` (to leave room for configurable defaults in a future update).
+
+#### Runtime Parameters
+
+Runtime parameters are provided as a mechanism for passing non-portable runtime-specific arguments
+that are generally static for the RuntimeClass, but dynamically interpreted by the runtime. They
+provide a way of specifying options that are tightly coupled with the class of runtime, but should
+not be used to enumerate general pod options as this would lead to a combinatorial explosion of
+RuntimeClass definitions.
+
+Examples of runtime parameters include:
+
+```
+// Decide which underlying OCI runtime should be used. This replaces the
+// annotations used by CRI-o, containerd & frakti to select between sandboxed
+// runtimes.
+"oci-runtime": "runsc"
+
+// Change the default set of capabilities used by the runtime. This is a good
+// way of surfacing behaviors that were previously implicit in docker but not
+// well defined in Kubernetes.
+"default-capabilities": "[]"
```

Comment:
Configuring defaults is eeking over towards general policy territory, but I see 2 reasons to include defaults in RuntimeClass parameters:

1. Our defaulting / policy story is a mess right now, and we don't have a good composable way of associating defaults with another field. In this case, setting defaults according to runtime class, irrespective of user.
2. Defaulting of fields otherwise not exposed in the API. If it's not configured on the pod spec, it can't be defaulted in the control plane.

Reply:
I like that this gives a mechanism for sane defaulting of capabilities. Similarly I wonder if this could be a mechanism for defaults for the other fields in the OCI runtime config - for example [hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-hooks), which afaict don't have any way to be configured at all right now)?

Reply:
I also think defaulting capabilities in `RuntimeClass` is practical approach: it does seem more like a runtime configuration than user specific.